drivers/serial/serial_mxc.c

```diff
@@ -318,10 +329,36 @@ static int mxc_serial_pending(struct udevice *dev, bool input)
 		return sr2 & USR2_TXDC ? 0 : 1;
 }
 
+static int barebox_mxc_serial_pending(struct udevice *dev, bool input)
+{
+	struct mxc_serial_plat *plat = dev_get_plat(dev);
+	struct mxc_uart *const uart = plat->reg;
+	uint32_t ts = readl(&uart->ts);
+
+	if (input)
+            return ts & UTS_RXEMPTY ? 0 : 1;
+	else
+            return ts & UTS_TXEMPTY ? 0 : 1;
+}
+static int barebox_mxc_serial_getc(struct udevice *dev)
+{
+	struct mxc_serial_plat *plat = dev_get_plat(dev);
+	struct mxc_uart *const uart = plat->reg;
+	while (readl(&uart->ts) & UTS_RXEMPTY);
+	return readl(&uart->rxd) & URXD_RX_DATA;
+}
+static int barebox_mxc_serial_putc(struct udevice *dev, const char ch)
+{
+	struct mxc_serial_plat *plat = dev_get_plat(dev);
+	struct mxc_uart *const uart = plat->reg;
+        while(readl(&uart->ts) & UTS_TXFULL);
+        writel(ch, &uart->txd);
+        return 0;
+}
 static const struct dm_serial_ops mxc_serial_ops = {
-	.putc = mxc_serial_putc,
-	.pending = mxc_serial_pending,
-	.getc = mxc_serial_getc,
+	.putc = barebox_mxc_serial_putc,
```

Comment:
Here you change the functionality of `putc` from non-blocking to blocking. I think that serial drivers in u-boot are designed non-blocking. This can have an influence on the "upper" layer.

Reply:
reverted back to non-blocking - with some minor patches to u-boots driver/serial/serial_mxc, which are beeing upstreamed